### PR TITLE
Mobile app settings backend

### PR DESCRIPTION
**Pull request by vadimkerr**
# What this PR does
Adds mobile app settings support to OnCall backend.

- Adds a new Django model `MobileAppUserSettings` to store push notification settings
- Adds a new endpoint `/mobile_app/v1/user_settings` to fetch/update settings from the mobile app

Some additional info on implementation: at first I wanted to extend the messaging backend system to allow storing / retrieving per-user data and implement mobile app settings based on those changes. After some thought I decided not to extend the messaging backend system and have this as functionality specific to the `mobile_app` Django app. Currently the messaging backend system is used by the backend and plugin UI, but mobile app settings are specific only to the mobile app and not configurable in the plugin UI.

**tldr: wanted to extend messaging backend system, but decided not to do that**

# Usage

## Get settings via API

`GET /mobile_app/v1/user_settings`
Example response:

```json
{
  "default_notification_sound_name": "default_sound",  # sound name without file extension
  "default_notification_volume_type": "constant",
  "default_notification_volume": 0.8,
  "default_notification_volume_override": false,
  "important_notification_sound_name": "default_sound_important",  # sound name without file extension
  "important_notification_volume_type": "constant",
  "important_notification_volume": 0.8,
  "important_notification_override_dnd": true
}
```

## Update settings via API
`PUT /mobile_app/v1/user_settings` - see example response above for payload shape.

Note that sound names must be passed without file extension. When sending push notifications, the backend will add `.mp3` to sound names and pass it to push notification data for Android. For iOS, sound names will be suffixed with `.aiff` to be used by APNS.

## Get settings from notification data for Android
All the settings from example response will be available in push notification data (along with `orgId`, `alertGroupId`, `title`, etc.).
Fields `default_notification_volume`, `default_notification_volume_override` and `important_notification_volume` , `important_notification_override_dnd` will be converted to strings due to FCM limitations.
Fields `default_notification_sound_name` and `important_notification_sound_name` will be suffixed with `.mp3` in push notification data.

## iOS limitations
While Android push notifications are handled purely on the mobile app side, iOS notifications are sent via APNS which imposes some limitations.
- Notification volume cannot be overridden for non-critical notifications (so fields `default_notification_volume_override` and `default_notification_volume` will be disregarded for iOS notifications)
- It's not possible to control volume type (i.e. "constant" vs "intensifying") via APNS. A possible workaround is to have different sound files for "constant" and "intensifying" and pass that as `default_notification_sound_name` / `important_notification_sound_name`.

# Which issue(s) this PR fixes
Related to https://github.com/grafana/oncall-private/issues/1602

# Checklist

- [x] Tests updated
- [x] No changelog updates since the changes are not user-facing

### CODE REVIEW

File: engine/apps/mobile_app/models.py [hunk 1]
```diff
@@ -68,3 +69,33 @@ def create_auth_token(cls, user: User, organization: Organization) -> Tuple["Mob
...
+
+    # APNS only allows to specify volume for critical notifications,
+    # so "default_notification_volume" and "default_notification_volume_override" are only used on Android
+    default_notification_volume = models.FloatField(
+        validators=[validators.MinValueValidator(0.0), validators.MaxValueValidator(1.0)], default=0.6
+    )
+    default_notification_volume_override = models.BooleanField(default=False)
+
+    # Push notification settings for critical notifications
+    critical_notification_sound_name = models.CharField(max_length=100, default="default")  # TODO: check sound name
```
Comment (vadimkerr) [inline engine/apps/mobile_app/models.py:95]: Default sound name TBD

**Review by matiasb — APPROVED** [review]
LGTM, one minor comment.

File: engine/apps/mobile_app/tasks.py [hunk 2]
```diff
@@ -87,8 +139,18 @@ def _create_error_log_record():
 
     alert_body = f"Status: {status_verbose}, alerts: {alerts_count_str}"
 
+    mobile_app_user_settings, _ = MobileAppUserSettings.objects.get_or_create(user=user)
+
+    # APNS only allows to specify volume for critical notifications
+    apns_volume = mobile_app_user_settings.critical_notification_volume if critical else None
```
[the remark this answers is not in the document]
Comment (vadimkerr) [reply engine/apps/mobile_app/tasks.py:145]: The volume setting here is passed to APNS which only applies to iOS, but `default_notification_volume` is only used by Android. I wanted to explicitly pass `None` here instead of `default_notification_volume` so it's clear that APNS doesn't take that parameter into account.

File: engine/apps/mobile_app/tasks.py [hunk 3]
```diff
@@ -123,48 +199,4 @@ def _create_error_log_record():
...
-    """
-    url = create_engine_url("mobile_app/v1/fcm_relay", override_base=settings.GRAFANA_CLOUD_ONCALL_API_URL)
-
-    response = requests.post(
-        url, headers={"Authorization": live_settings.GRAFANA_CLOUD_ONCALL_TOKEN}, json=json.loads(str(message))
-    )
-    response.raise_for_status()
-
-    return response
+    return message
```
Comment (joeyorlando) [inline engine/apps/mobile_app/tasks.py:202]: _nit_
```suggestion
```

File: engine/apps/mobile_app/tasks.py [hunk 4]
```diff
@@ -87,8 +139,18 @@ def _create_error_log_record():
...
+
+    # APNS only allows to specify volume for critical notifications
+    apns_volume = mobile_app_user_settings.critical_notification_volume if critical else None
+    apns_sound_name = (
+        mobile_app_user_settings.critical_notification_sound_name
+        if critical
+        else mobile_app_user_settings.default_notification_sound_name
+    )
+
     message = Message(
```
Comment (joeyorlando) [inline engine/apps/mobile_app/tasks.py:152]: _nit_
```suggestion
    return Message(
```

**Review by joeyorlando — APPROVED** [review]
lgtm! I'm sure people will be pleased once this is released 😛

**Review by Dieterbe — APPROVED** [review]
since you requested a review, i can say that it looks fine in principle, but obviously i didn' treview the python code.

File: engine/apps/mobile_app/views.py [hunk 5]
```diff
@@ -50,3 +52,13 @@ def delete(self, request):
...
         return Response(status=status.HTTP_204_NO_CONTENT)
+
+
+class MobileAppUserSettingsAPIView(generics.RetrieveUpdateAPIView):
+    authentication_classes = (MobileAppAuthTokenAuthentication,)
+    permission_classes = (IsAuthenticated,)
+    serializer_class = MobileAppUserSettingsSerializer
+
+    def get_object(self):
+        mobile_app_settings, _ = MobileAppUserSettings.objects.get_or_create(user=self.request.user)
```
[the remark this answers is not in the document]
Comment (vadimkerr) [reply engine/apps/mobile_app/views.py:63]: Yes, the first time the endpoint is called or the first time user is to be notified with a push notification
